Fix for auth version change during node startup

nuivall opened the pull request:
Before this patch we may trigger `SCYLLA_ASSERT(legacy_mode(_qp))`.
That's because some auth startup is done in the background
and assumes that auth version doesn't change in the middle
of the startup. But topology coordinator may decide to do
the migration at any time, regadless if auth service is
fully started on all nodes.

This change makes sure that in legacy startup flow we'll
always use old auth-v1 keyspace and therefore auth version
change in the middle won't negatively affect the flow.

Fixes https://github.com/scylladb/scylladb/issues/25505

andrzej-jackowski-scylladb reviewed (approved):
Minor comment regarding #include.
The change looks ok.

piodul reviewed (changes requested):
I am not sure the check is 100% solid. The `standard_role_manager::start` function, which calls the problematic `legacy_create_default_role_if_missing` function with the assert looks like this:

```c++
future<> standard_role_manager::start() {
    return once_among_shards([this] () -> future<> {
        if (legacy_mode(_qp)) {
            co_await create_legacy_metadata_tables_if_missing();
        }

        auto handler = [this] () -> future<> {
            const bool legacy = legacy_mode(_qp);
            if (legacy) {
                if (!_superuser_created_promise.available()) {
                    _superuser_created_promise.set_value();
                }
                co_await _migration_manager.wait_for_schema_agreement(_qp.db().real_database(), db::timeout_clock::time_point::max(), &_as);

                if (co_await any_nondefault_role_row_satisfies(_qp, &has_can_login)) {
                    if (legacy_metadata_exists()) {
                        log.warn("Ignoring legacy user metadata since nondefault roles already exist.");
                    }
                    co_return;
                }

                if (legacy_metadata_exists()) {
                    co_await migrate_legacy_metadata();
                    co_return;
                }
                co_await legacy_create_default_role_if_missing();
            }
            if (!legacy) {
                co_await maybe_create_default_role_with_retries();
                if (!_superuser_created_promise.available()) {
                    _superuser_created_promise.set_value();
                }
            }
        };

        _stopped = auth::do_after_system_ready(_as, handler);
        co_return;
    });
}
```

The `_superuser_created_promise` promise - for some reason that is absolutely unknown to me - is set _before_ the superuser is actually created, if we are in legacy mode. Between this and actually trying to create the superuser, there is the schema agreement wait - which should make the race more likely.

I believe that the race can still happen until we actually wait that we got past the`legacy_create_default_role_if_missing` call.

There is an analogous problem with the `password_authenticator` object.

nuivall commented:
v2:
- added callback instead of auth dependency
- added abort_source
- changed when promise is ready

piodul commented on auth/password_authenticator.cc:
```diff
@@ -209,33 +209,34 @@ future<> password_authenticator::start() {
-                        _superuser_created_promise.set_value();
-                    }
                 }
+                ready();
```
Nit: I've got a suggestion on how to do it in a cleaner way: add a continuation to the future returned by `async` where you conditionally set the `_superuser_created_promise`. This way, it will be handled in one place and nobody will have to remember to add further `ready` calls (unlikely, but also makes it easier to read the code, at least in my opinion).

piodul commented on auth/standard_role_manager.cc:
```diff
@@ -303,31 +303,33 @@ future<> standard_role_manager::start() {
+            ready();
         };
 
         _stopped = auth::do_after_system_ready(_as, handler);
```
Not sure how you could handle this case with respect to `_superuser_created_promise`, so perhaps it's better to leave it like this and also keep the `password_authenticator` code as it is now as well.

piodul reviewed (commented):
I've left a single nitpick of a dubious value as a comment to the code.

One question before leaving a verdict. I am worried that you might run into a situation where a non-coordinator node is restarted (e.g. it is upgraded) and then the topology coordinator decides to run the upgrade, before the restarted node finished setting up the auth service. What will happen in that case? Do I understand correctly that if the superuser is already created then it is safe to run the upgrade in parallel? If so then I think that the current approach __will__ work, but it's not immediately obvious - it would be good to write it down in the PR description.

A regression test would be nice to have - I think you can write one relatively easily for the scenario where the restarted node != topology coordinator (which I think is a omre important case than when restarted node == topology coordinator).

nuivall commented:
> The _superuser_created_promise promise - for some reason that is absolutely unknown to me - is set before the superuser is actually created

I finally found the reason for this. It would be easier to spot if the dependencies were managed more explicitly. There is wait_for_schema_agreement on legacy startup path, without any timeout, and it means auth-v1 can't start without majority, that's why we didn't block on auth-v1 startup.

nuivall committed:
auth: document setting _superuser_created_promise flow in auth-v1

nuivall committed:
auth: use old keyspace during auth-v1 consistently

Before this patch we may trigger assertion on legacy_mode(_qp).
That's because some auth startup is done in the background
and assumes that auth version doesn't change in the middle
of the startup. But topology coordinator may decide to do
the migration at any time, regadless if auth service is
fully started on all nodes.

This change makes sure that in legacy startup flow we'll
always use old auth-v1 keyspace and therefore auth version
change in the middle won't negatively affect the flow.

nuivall committed:
auth: mark some auth-v1 functions as legacy

piodul commented on auth/password_authenticator.cc:
```diff
@@ -119,8 +123,7 @@ future<> password_authenticator::migrate_legacy_metadata() const {
 }
 
 future<> password_authenticator::legacy_create_default_if_missing() {
```
I think you missed the `update_row_query` call in this function.

nuivall replied:
yes, thanks, I'll move string directly to this func, little benefit in adding param to this function

piodul commented on auth/password_authenticator.cc:
```diff
@@ -119,7 +123,6 @@ future<> password_authenticator::migrate_legacy_metadata() const {
 }
 
 future<> password_authenticator::legacy_create_default_if_missing() {
```
`maybe_create_default_password_with_retries` is also called during the `do_after_system_ready` flow, maybe it should be updated as well?

Similarly for `maybe_create_default_role_with_retries` in `standard_role_manager`.

nuivall replied:
yes it is, but this is auth-v2 flow, since there is no v2->v1 downgrade, and we don't know how potential future v2->v3 would look like, we don't need to change it

nuivall commented:
v4: fixed update_row_query in legacy_create_default_if_missing
